### PR TITLE
ENH Add Poisson datafit

### LINKED ISSUE #62: ENH add Poisson datafit

### PR DESCRIPTION
**Pull request by PABannier**
Closes #62

### CODE REVIEW

**Comment by PABannier** [comment]
Note to self: Poisson datafit is not Lipschitz, need for a line search

**Review by Badr-MOUFAD — COMMENTED** [review]
You can instead use PN solver as it doesn't require datafits to be Lipschitz and performs a backtracking line search to find a suitable step.
We should implement two functions ``raw_gradient`` and ``raw_hessian`` as in the ``Logisitc`` datafit

https://github.com/scikit-learn-contrib/skglm/blob/19a5a6f52200cf3eca7eb6f5e3f342d50ead39c3/skglm/datafits/single_task.py#L132-L139

Did I get what you meant @PABannier?

**Comment by mathurinm** [comment]
Casting `y` to positive values, I get convergence for all designs I tried (even 5000 x 3000). I think @PABannier you encountered issues because for negative ys, the loss is no longer convex.

It remains to add an example, test against `sklearn.linear_model.PoissonRegressor` and we are good to go on my side

File: test_poisson.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import numpy as np
```
Comment (Badr-MOUFAD) [inline test_poisson.py:1]: ``statsmodels`` provides $\ell_1$ regulized  [Poisson regressor](https://www.statsmodels.org/dev/generated/statsmodels.discrete.discrete_model.Poisson.fit_regularized.html#statsmodels.discrete.discrete_model.Poisson.fit_regularized).
We better unittest against them.

**Comment by mathurinm** [comment]
Good idea ! But we don't want statsmodel as a dependency for the main code, so it will need to be a test dependency only 

I thought sklearn supported regularization, but it's only L2 apparently. 

To discuss: it we match sklearn on Poisson + L2, is it enough or do we need to add a statsmodel test requirement to test Poisson + L1 ? @QB3 @Klopfe @PABannier thoughts on this tradeoff ?

**Comment by PABannier** [comment]
I'd opt for comparing l1-regularized models. I've added `statsmodels` as a test dependency.

**Comment by PABannier** [comment]
For the example, do you think about something in particular @mathurinm ?

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -12,17 +12,18 @@ jobs:
...
-    - name: Install dependencies
-      run: |
-        python -m pip install --upgrade pip
-        python -m pip install git+https://github.com/mathurinm/celer.git
-        pip install pytest
-        pip install numpydoc
-        pip install .
-    - name: Test with pytest
-      run: pytest -v skglm/
+      - uses: actions/checkout@v2
```
Comment (mathurinm) [inline .github/workflows/main.yml:15]: @PABannier can you revert this big chunk of changes ? Normally there's just one line that changed

File: skglm/tests/test_datafits.py [hunk 3]
```diff
@@ -3,12 +3,13 @@
 
 from sklearn.linear_model import HuberRegressor
 from numpy.testing import assert_allclose, assert_array_less
+from statsmodels.discrete.discrete_model import Poisson as PoissonRegressor
```
Comment (mathurinm) [inline skglm/tests/test_datafits.py:6]: See discussion here: https://github.com/scikit-learn-contrib/skglm/pull/57/files#r990803934

we need a more advanced scheme to avoid imposing statsmodel for every user, yet to make it clear that it's a test dependency

**Review by mathurinm — CHANGES_REQUESTED** [review]
PR looks great, once again a nice show of force from our design and solvers: it costs 20 lines to add a whole new model !!!

We only need to be a bit more careful with the statsmodel dependecy and that will be a green light for me

**Comment by mathurinm** [comment]
> For the example, do you think about something in particular @mathurinm ?

It can be anything really, but to quote the great man "if it's not in the doc, it does not exist". Maybe check out what sklearn has ?

File: skglm/tests/test_datafits.py (hunk 3 above)
Comment (PABannier) [reply skglm/tests/test_datafits.py]: i've changed it

File: skglm/tests/test_datafits.py [hunk 4]
```diff
@@ -56,5 +56,37 @@ def test_log_datafit():
     np.testing.assert_almost_equal(-grad * (y + n_samples * grad), hess)
 
 
+def test_poisson():
+    try:
+        from statsmodels.discrete.discrete_model import Poisson as PoissonRegressor  # noqa
+    except ImportError:
+        pytest.xfail()
```
Comment (mathurinm) [inline skglm/tests/test_datafits.py:63]: Great! 
Nitpick: can you add an error message?

**Review by mathurinm — APPROVED** [review]

**Comment by mathurinm** [comment]
Thanks @PABannier
